btcjson/walletsvrcmds.go

```diff
@@ -686,6 +692,105 @@ func NewWalletPassphraseChangeCmd(oldPassphrase, newPassphrase string) *WalletPa
+type PsbtInput struct {
+	Txid     string `json:"txid"`
+	Vout     uint32 `json:"vout"`
+	Sequence uint32 `json:"sequence"`
+}
+
+// PsbtOutput represents an output to include in the PSBT created by the
+// WalletCreateFundedPsbtCmd command.
+type PsbtOutput map[string]interface{}
+
+// NewPsbtOutput returns a new instance of a PSBT output to use with the
+// WalletCreateFundedPsbtCmd command.
+func NewPsbtOutput(address string, amount btcutil.Amount) PsbtOutput {
+	out := make(map[string]interface{})
+	out[address] = amount.ToBTC()
+	return out
+}
+
+// NewPsbtDataOutput returns a new instance of a PSBT data output to use with
+// the WalletCreateFundedPsbtCmd command.
+func NewPsbtDataOutput(data []byte) PsbtOutput {
+	out := make(map[string]interface{})
+	out["data"] = hex.EncodeToString(data)
+	return out
+}
+
+// WalletCreateFundedPsbtOpts represents the optional options struct provided
+// with a WalletCreateFundedPsbtCmd command.
+type WalletCreateFundedPsbtOpts struct {
+	ChangeAddress          *string     `json:"changeAddress,omitempty"`
+	ChangePosition         *int64      `json:"changePosition,omitempty"`
+	ChangeType             *ChangeType `json:"change_type,omitempty"`
+	IncludeWatching        *bool       `json:"includeWatching,omitempty"`
+	LockUnspents           *bool       `json:"lockUnspents,omitempty"`
+	FeeRate                *int64      `json:"feeRate,omitempty"`
+	SubtractFeeFromOutputs *[]int64    `json:"subtractFeeFromOutputs,omitempty"`
+	Replaceable            *bool       `json:"replaceable,omitempty"`
+	ConfTarget             *int64      `json:"conf_target,omitempty"`
+	EstimateMode           *string     `json:"estimate_mode,omitempty"`
+}
+
+// WalletCreateFundedPsbtCmd defines the walletcreatefundedpsbt JSON-RPC command.
+type WalletCreateFundedPsbtCmd struct {
+	Inputs      []PsbtInput
+	Outputs     []PsbtOutput
+	Locktime    *uint32
+	Options     *WalletCreateFundedPsbtOpts
+	Bip32Derivs *bool
+}
+
+// NewWalletCreateFundedPsbtCmd returns a new instance which can be used to issue a
+// walletcreatefundedpsbt JSON-RPC command.
+func NewWalletCreateFundedPsbtCmd(
+	inputs []PsbtInput, outputs []PsbtOutput, locktime *uint32,
+	options *WalletCreateFundedPsbtOpts, bip32Derivs *bool,
+) *WalletCreateFundedPsbtCmd {
+	return &WalletCreateFundedPsbtCmd{
+		Inputs:      inputs,
+		Outputs:     outputs,
+		Locktime:    locktime,
+		Options:     options,
+		Bip32Derivs: bip32Derivs,
+	}
+}
+
+// SighashType defines the different signature hash types to sign with if not
+// specified by the PSBT, for use with the walletprocesspsbt JSON-RPC command.
+type SighashType string
+
+var (
+	SighashAll                SighashType = "ALL"
+	SighashNone               SighashType = "NONE"
+	SighashSingle             SighashType = "SINGLE"
+	SighashAllAnyoneCanPay    SighashType = "ALL|ANYONECANPAY"
+	SighashNoneAnyoneCanPay   SighashType = "NONE|ANYONECANPAY"
+	SighashSingleAnyoneCanPay SighashType = "SINGLE|ANYONECANPAY"
+)
```

Comment:
Would be nice to have comments here as well.

Reply:
Actually, I just noticed that rpcclient already defines a `SigHashType` type. I think it will be more clear to use that and then convert to `*string` for the btcjson constructors (to avoid a circular import). This is consistent with the behaviour of `SignRawTransaction`.

Reply:
`SigHashType` should've really been in `btcjson` package. But there's nothing we can do about it now, without introducing a breaking change. Let's keep it for later.